### PR TITLE
SubwordVocab interface and FinalfusionBucketVocab.

### PR DESCRIPTION
**Pull request by sebpuetz**
Define SubwordVocab interface and implement FinalfusionBucketVocab.
Remove the option to pass a word-index mapping to vocabularies.

### CODE REVIEW

File: src/finalfusion/vocab/subword.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    """
+    Interface for vocabularies with subword lookups.
+    """
+    def idx(self, item: str, default: Optional[Union[List[int], int]] = None
+            ) -> Optional[Union[List[int], int]]:
+        idx = self.word_index.get(item)
+        if idx is not None:
+            return idx
+        subwords = cast(List[int], self.subword_indices(item))
+        if subwords:
```
Comment (danieldk) [inline src/finalfusion/vocab/subword.py]: I know that this is valid Python, but I think the code is easier to read if this is actually a concrete check against a value (if it is not a boolean).

File: src/finalfusion/vocab/subword.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                                                    offset=len(self.words),
+                                                    bracket=bracket,
+                                                    with_ngrams=with_ngrams)
+
+    def __getitem__(self, item: str) -> Union[int, List[int]]:
+        idx = self.word_index.get(item)
+        if idx is not None:
+            return idx
+        subwords = cast(List[int], self.subword_indices(item))
+        if subwords:
```
Comment (danieldk) [inline src/finalfusion/vocab/subword.py]: See above.

File: src/finalfusion/vocab/subword.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+            return subwords
+        raise KeyError(f"No indices found for {item}")
+
+    def __eq__(self, other: Any) -> bool:
+        return super(SubwordVocab, self).__eq__(other) \
```
Comment (danieldk) [inline src/finalfusion/vocab/subword.py]: For hashing subword vocabs comparing the indexers is faster I'd assume. Maybe swap the order? Or defer the implementation to concrete implementations?

**Review by danieldk — CHANGES_REQUESTED** [review]

File: src/finalfusion/vocab/subword.py (hunk 3 above)
Comment (sebpuetz) [reply src/finalfusion/vocab/subword.py]: Simply changing order will break comparison to `SimpleVocab`s, `Vocab.__eq__` has the type check which short-circuits if the types are different. `SubwordVocab.__eq__` accesses `subword_indexer` which raises an `AttributeError` for `SimpleVocab`.

I added an additional type check in `SubwordVocab.__eq__` and changed the order of comparison. There shouldn't be much of a difference when comparing the explicit indexers first vs. the vocabulary parts, but the improvement for hash indexers should be quite large.

**Comment by sebpuetz** [comment]
Thanks for the review, I pushed the revised version!

File: src/finalfusion/vocab/subword.py (hunk 1 above)
Comment (sebpuetz) [reply src/finalfusion/vocab/subword.py]: PyCharm complains but pylint stays quite, so I'm okay with comparing to a list literal.

**Review by danieldk — APPROVED** [review]
Looks good!

**Pull request merged**